src/ui/ui_review_menu_nbgl.c

```diff
@@ -0,0 +1,335 @@
+} nbgl_tx_infos_t;
+
+// Static variables
+static nbgl_layoutTagValueList_t pairList;
+static nbgl_pageInfoLongPress_t infoLongPress;
+static nbgl_tx_infos_t txInfos;
+
+// Static functions declarations
+static void prepareTxInfos(ui_approval_state_t state, bool data_warning);
+static void reviewStart(void);
+static void displayTransaction(void);
+static void displayDataWarning(void);
+static void displayCustomContractWarning(void);
+static void dataWarningChoice(bool reject);
+static void customContractWarningChoice(bool reject);
+static void reviewChoice(bool confirm);
+static void rejectConfirmation(void);
+static void rejectChoice(void);
+
+static void dataWarningChoice(bool accept) {
+    if (accept) {
+        if (txInfos.warnings[CUSTOM_CONTRACT_WARNING] == true) {
+            displayCustomContractWarning();
+        } else {
+            displayTransaction();
+        }
+    } else {
+        txInfos.rejectCb();
+        nbgl_useCaseStatus("Transaction rejected", false, ui_idle);
+    }
+}
+
+static void customContractWarningChoice(bool accept) {
+    if (accept) {
+        displayTransaction();
+    } else {
+        txInfos.rejectCb();
+        nbgl_useCaseStatus("Transaction rejected", false, ui_idle);
+    }
+}
+
+static void displayDataWarning(void) {
+    nbgl_useCaseChoice(&C_round_warning_64px,
+                       "WARNING\nData Present\nProceed with care",
+                       "Reject if you're not sure",
+                       "Continue",
+                       "Reject",
+                       dataWarningChoice);
+}
+
+static void displayCustomContractWarning(void) {
+    nbgl_useCaseChoice(&C_round_warning_64px,
+                       "WARNING\nCustom Contract\nProceed with care",
+                       "Reject if you're not sure",
+                       "Continue",
+                       "Reject",
+                       customContractWarningChoice);
+}
+
+static void displayTransaction(void) {
+    nbgl_useCaseStaticReview(&pairList, &infoLongPress, "Reject", reviewChoice);
+}
+
+static void reviewStart() {
+    nbgl_callback_t displayFunction = displayTransaction;
+    if (txInfos.warnings[DATA_WARNING] == true) {
+        displayFunction = displayDataWarning;
+    } else if (txInfos.warnings[CUSTOM_CONTRACT_WARNING] == true) {
+        displayFunction = displayCustomContractWarning;
+    }
+    nbgl_useCaseReviewStart(&C_stax_app_tron_64px,
+                            txInfos.flowTitle,
+                            txInfos.flowSubtitle,
+                            "Reject",
+                            displayFunction,
+                            rejectChoice);
+}
+
+static void reviewChoice(bool confirm) {
+    if (confirm) {
+        nbgl_useCaseStatus("TRANSACTION\nCONFIRMED", true, ui_idle);
+        txInfos.confirmCb();
+    } else {
+        rejectChoice();
+    }
+}
+
+static void rejectConfirmation(void) {
+    nbgl_useCaseStatus("Transaction rejected", false, ui_idle);
+    txInfos.rejectCb();
```

Comment:
Sometimes you display the status then call the callback, and sometimes you do it the other way (in `customContractWarningChoice` for example).
Can you harmonize it?

Reply:
Ok